hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,717 @@
+    addrBuf[0] = (uint8_t)((address>>16) & 0xFF);
+    addrBuf[1] = (uint8_t)((address>>8) & 0xFF);
+    addrBuf[2] = (uint8_t)((address>>0) & 0xFF);
+
+    nrfx_err_t ret = nrfx_qspi_lfm_start(&writeSecurityRegister);
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+    nrf_delay_us(10); // need a bit delay to switch between normal flash and security register
+    ret = nrfx_qspi_lfm_xfer(addrBuf, nullptr, sizeof(addrBuf), false); // Send address bytes, do not finalize long frame transfer
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+    ret = nrfx_qspi_lfm_xfer(buf, nullptr, size, true); // Send data to write, finalize long frame transfer
+    CHECK_TRUE(ret == NRFX_SUCCESS, nrf_system_error(ret));
+
+    return 0;
+}
+
+int erase_common(uintptr_t start_addr, size_t num_blocks, nrf_qspi_erase_len_t len) {
+    FlashLock lk;
+
+    int err_code = NRF_SUCCESS;
+
+    const size_t block_length = len == NRF_QSPI_ERASE_LEN_4KB ? 4096 : 64 * 1024;
+
+    /* Round down to the nearest block */
+    start_addr = ((start_addr / block_length) * block_length);
+
+    for (size_t i = 0; i < num_blocks; i++) {
+        err_code = nrfx_qspi_erase(len, start_addr);
+        CHECK_TRUE(err_code == NRFX_SUCCESS, nrf_system_error(err_code));
+
+        exflash_qspi_wait_completion();
+
+        start_addr += block_length;
+    }
+
+    // LOG_DEBUG(TRACE, "Erased %lu %lukB blocks starting from %" PRIxPTR,
+    //           num_blocks, block_length / 1024, start_addr);
+
+    return 0;
+}
+
+} // Anonymous
+
+int hal_exflash_init(void) {
+    // PATCH: Initialize CS pin, external memory discharge
+    nrf_gpio_cfg_output(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_pin_set(QSPI_FLASH_CSN_PIN);
+    nrf_gpio_cfg_input(QSPI_FLASH_IO1_PIN, NRF_GPIO_PIN_PULLDOWN);
+
+    FlashLock lk;
+
+    // Use single data line SPI to read ID
+    nrfx_qspi_config_t default_config = NRF_QSPI_QUICK_CFG(NRF_QSPI_WRITEOC_PP, NRF_QSPI_READOC_FASTREAD, NRF_QSPI_FREQ_32MDIV2);
+    int ret = nrfx_qspi_init(&default_config, nullptr, nullptr);
+    if (ret) {
+        if (ret == NRFX_ERROR_TIMEOUT) {
+            // Could have timed out because the flash chip might be in an ongoing program/erase operation.
+            // Suspend it.
+            LOG_DEBUG(TRACE, "QSPI NRFX timeout error. Suspend pgm/ers op.");
+            ret = hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SUSPEND_PGMERS, nullptr, nullptr, 0);
+        }
+        return nrf_system_error(ret);
+    }
+
+    // Determine chip type / wake up from potential sleep
+    uint8_t chip_id[3] = {};
+    nrf_qspi_cinstr_conf_t read_id_command = {
+        .opcode    = HAL_QSPI_CMD_STD_READ_ID,
+        .length    = NRF_QSPI_CINSTR_LEN_4B,
+        .io2_level = true,
+        .io3_level = true,
+        .wipwait   = true,
+        .wren      = true
+    };
+    
+    // If read chip_id fails, it may be a GD25 that is asleep
```

Comment:
Why not switch type (`flash_type`) to GD25 temporarily and delegate to `hal_exflash_special_command` ?

Reply:
@scott-brust Read ID should be supported in the Deep Power-Down Mode.

>[GD25Q32C Datasheet]
>The Deep Power-Down Mode can only be entered by executing the Deep Power-Down (DP) command. Once the device has entered the Deep Power-Down Mode, all commands are ignored except the Release from Deep Power-Down and Read Device ID (RDI) (ABH) or Enable Reset (66H) and Reset (99H) commands.

By the way, I would consider that the read status command(RDSR), which is used in `nrfx_qspi_init()` would fail in Deep Power-Down Mode.